template/gitlab-runner.tftpl

```diff
@@ -260,6 +260,134 @@ then
   yum install gitlab-runner-${gitlab_runner_version} -y
 fi
 
+#############################
+# GRACEFUL SHUTDOWN SERVICE #
+#############################
+
+# set timeout for runner service to wait before stopping
+mkdir /etc/systemd/system/gitlab-runner.service.d
+cat <<EOF > /etc/systemd/system/gitlab-runner.service.d/kill.conf
+[Service]
+# This initiates a graceful shutdown of the service: https://docs.gitlab.com/runner/commands/#signals
+KillSignal=SIGQUIT
+TimeoutStopSec=0
+EOF
+
+cat <<'EOF' > /opt/monitor_runner.sh
+#!/bin/bash
+
+# Fetch instance ID since we need it for the lifecycle hook actions.
+imds_token=$(curl -sSf -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 300")
+instance_id=$(curl -sS -H "X-aws-ec2-metadata-token: $imds_token" "http://169.254.169.254/latest/meta-data/instance-id")
+region=$(curl -s -H "X-aws-ec2-metadata-token: $imds_token" "http://169.254.169.254/latest/meta-data/placement/region")
+
+# Function to send complete-lifecycle-action
+send_complete_lifecycle_action() {
+    echo "Sending complete-lifecycle-action for instance $instance_id..."
+
+    if [ -z "$region" ]; then
+    echo "Failed to retrieve AWS region."
+    exit 1
+    fi
+    echo "Current AWS Region: $region"
+
+    # Retrieve the Auto Scaling Group name associated with the instance
+    asg_name=$(aws --region $region ec2 describe-tags --filters "Name=resource-id,Values=$instance_id" "Name=key,Values=aws:autoscaling:groupName" --query 'Tags[].Value' --output text)
+    if [ -z "$asg_name" ]; then
+    echo "Failed to retrieve Auto Scaling Group name."
+    exit 1
+    fi
+    echo "Auto Scaling Group Name: $asg_name"
+
+    # Retrieve the lifecycle hook names associated with the Auto Scaling Group
+    lifecycle_hooks=$(aws --region $region autoscaling describe-lifecycle-hooks --auto-scaling-group-name $asg_name --query "LifecycleHooks[*].LifecycleHookName" --output text)
```

Comment:
Do you think it'd be safer to filter to only "instance terminating" hooks? I was thinking it's possible that some users have "instance launching" hooks attached to the ASG.

Reply:
Thanks for the suggestion! Updated the query.